Implement File block

noisysocks opened the pull request:
Based off of @mirka's work in https://github.com/WordPress/gutenberg/pull/6805.

Closes #5462.

Implements a File block with the following features:

- Editable text link
- Text link can be copy and pasted to create a normal link
- HTML5 download link
- Copy URL button to copy the media file URL to clipboard
- Accepts both Media Library files and external URLs only Media Library files
- Text link can be changed to point to the Attachment Page
- Text link can be changed to open in a new window
- Accepts and uploads a file drag-and-dropped into the editor (or placeholder)
- Transforms to/from Audio and Video blocks
- Download button text is editable
- Download button is optional

noisysocks committed:
Implement File block

Implements a File block with the following features:

- Editable text link
- Text link can be copy and pasted to create a normal link
- HTML5 download link
- Copy URL button to copy the media file URL to clipboard
- Accepts both Media Library files and external URLs only Media Library files
- Text link can be changed to point to the Attachment Page
- Text link can be changed to open in a new window
- Accepts and uploads a file drag-and-dropped into the editor (or placeholder)
- Transforms to/from Audio and Video blocks
- Download button text is editable
- Download button is optional

Co-authored-by: mirka <mirka@jaguchi.com>

noisysocks committed:
e → event

noisysocks committed:
Specify type on all of the file block's attributes

noisysocks committed:
Use  instead of calling setAttributes() in constructor

noisysocks committed:
this.props.attributes → attributes

noisysocks committed:
Use classnames to simplify className computation

noisysocks committed:
this.props.hrefs → hrefs

noisysocks committed:
Use let instead of IIFE

noisysocks committed:
Add help text to Show Download Button toggle, improve copy

noisysocks committed:
Make FileBlockInspector a function component

talldan committed:
Move inline functions in render to class methods

talldan committed:
Minor stylistic changes in editable-link component

- Move inline functions in render method to class methods
- Change `e` to `event`
- Remove unnecessary if statement

talldan committed:
Add visited styling for file block button in front-end

talldan committed:
Add explanatory comment about use of RichText component

talldan committed:
Renaming and commenting for file block attributes

talldan committed:
Renaming for openInNewWindow attribute for clarity

talldan commented:
It seems possible to transform file blocks that are not audio or video to those block types (e.g. a pdf).

noisysocks committed:
Fix file block content tests

noisysocks commented:
This is in pretty good shape now. @talldan and reviewed and tested the code together, and fixed up some minor issues. Aiming to merge this early next week.

Soean commented:
Nice work! What do you think about selecting/changing the color of the download button?

talldan commented on core-blocks/file/index.js:
```diff
@@ -0,0 +1,164 @@
+				transform: ( attributes ) => {
+					return createBlock( 'core/file', {
+						href: attributes.src,
+						fileName: attributes.caption && attributes.caption.join(),
```
Created issue #7626 for this.

mcsf commented on core-blocks/file/index.js:
```diff
@@ -0,0 +1,172 @@
+				},
+			},
+		],
+		to: [
```
It's pretty strange to unconditionally offer transforms to audio and video regardless of file type, but I also understand that there's no matching mechanism in transforms (unlike `isMatch` when converting classic content to blocks). We could tackle this later, but just noting that the UX can be a bit weird (offering a transform that will break if the file is, say, a ZIP file).

talldan replied:
Yep, agree with that. I've created an issue for this - #7625.

The question is whether we're happy to leave it as it is and merge this, or whether we want to remove the transforms temporarily.

Another alternative could be to show a warning on audio and video blocks if the file type is incorrect.

noisysocks replied:
I've no strong opinion either way. Leaning slightly to leaving it as is for now because if we fix the bug in a future PR it aligns better with our goal of having the next release be 'feature complete'.

mcsf commented on editor/utils/editor-media-upload/index.js:
```diff
@@ -14,7 +14,7 @@ import { mediaUpload } from '@wordpress/utils';
  *
  * @param   {Object}   $0                   Parameters object passed to the function.
- * @param   {string}   $0.allowedType       The type of media that can be uploaded.
+ * @param   {string}   $0.allowedType       The type of media that can be uploaded, or '*' to allow all.
```
Is it better to require an explicit `'*'` or to make `allowedType` optional and assume no restriction? (I'm not sure)

talldan replied:
I had a think about it, and I think I prefer the `*` over making it optional. It at least signals a clear intention and is an accepted wildcard.

Making it optional and leaving it undefined could indicate either any or no allowed types.

Another option is to use a constant for the accepted values - e.g `allowedType: fileTypes.image` or `allowedType: fileTypes.any`.

mcsf commented on core-blocks/file/index.js:
```diff
@@ -0,0 +1,172 @@
+						href: blobURL,
+						fileName: file.name,
+						textLinkHref: blobURL,
+					} );
```
Here's what I got when dragging a small ZIP file into the editor:

<img width="700" alt="screen shot 2018-06-29 at 14 56 23" src="https://user-images.githubusercontent.com/150562/42097911-4e0b90fe-7bb1-11e8-8c79-c72b21d4ad5f.png">

talldan replied:
Fixed in 5202a06

mcsf reviewed (commented):
This looks great, even though I left some generally secondary comments.

The `'file'` transform I'd have a better look at, either fixing or removing it.

talldan committed:
Use wp.blob instead of deprecated  wp.utils

talldan committed:
Fix error with file block when dragging/dropping file

When dragging and dropping a file to create a new block an error
was being thrown. This is caused by attempting to use `getBlobByUrl`
as though it returns a promise. This function seems to be synchronous
so the fix is to use it that way.

talldan committed:
Improve copy for file block description

noisysocks commented:
Thanks for addressing those comments @talldan. This looks good to me! 👍:shipit:

pento reviewed (approved):
Looks good to me, let's ship it!

Fun side note, I was initially running into max file size issues, which I thought had been fixed. We really should get #4203 merged, too.

noisysocks commented:
Thanks @pento! And props again to @mirka for the great work here 🎉🌈💃✨

talldan commented:
and also thanks @mcsf for the helpful review comments 👍

aduth commented on core-blocks/file/index.js:
```diff
@@ -0,0 +1,172 @@
+	transforms: {
+		from: [
+			{
+				type: 'files',
```
Seems to me this should have been given a `priority` of low value, since as implemented it has identical priority as any other file transform, when in fact this should be the fallback. For example, dropping an image should take advantage of the Image block's `files` transform before this one. I don't know if this is currently working, and if it is it only incidental and prone to future breakage.

See: https://wordpress.org/gutenberg/handbook/block-api/#transforms-optional

>To control the priority with which a transform is applied, define a priority numeric property on your transform object, where a lower value will take precedence over higher values. This behaves much like a WordPress hook. Like hooks, the default priority is 10 when not otherwise set.

aduth replied:
Issue at #8022

aduth commented on core-blocks/file/editable-link.js:
```diff
@@ -0,0 +1,79 @@
+					onCopy={ this.copyLinkToClipboard }
+					onCut={ this.copyLinkToClipboard }
+					onPaste={ this.forcePlainTextPaste }
+					contentEditable
```
Warning:

>Warning: A component is `contentEditable` and contains `children` managed by React. It is now your responsibility to guarantee that none of those nodes are unexpectedly modified or duplicated. This is probably not intentional.

Should this be using the `RichText` component instead? There are many complexities of `contentEditable` that are meant to be abstracted into this component for general usage like this.

aduth commented:
This block appears to completely fail to handle file types not supported for upload. See #8025.
